[PLT-1015] Make all Mattermost system emails Multi-Part, with both a HTML and Plain Text version.

cpanato opened the pull request:
#### Summary
This PR make all emails sent by mattermost both text and HTML.
Please let me know if the library used to convert HTML to text is good. Otherwise, we will need to write the text templates as same we have for the HTML.

Added in WIP for review because if we need to use text templates I will close this PR and implement that.

#### Ticket Link
Jira: https://mattermost.atlassian.net/browse/PLT-1015
GH: https://github.com/mattermost/platform/issues/5486

#### Checklist
- [x] Added or updated unit tests

jasonblais commented on glide.yaml:
```diff
@@ -79,3 +79,6 @@ import:
 - package: github.com/dyatlov/go-opengraph
   subpackages:
   - opengraph
+- package: github.com/jaytaylor/html2text
```
both of these licenses look good

jasonblais commented:
@grundleborg maybe help take an initial look at the code before we proceed?

cpanato commented:
if this stuff i did for the glide is the correct one, I will squash the commits

coreyhulen commented:
Maybe we can have @crspeller add the libs needed then you can submit this change afterwards since we'll upgrade the libs soon after the 3.7 release.

That way it'll only be the diff of the MM code change.

cpanato commented:
ok @coreyhulen let me know when this is ready then I can rebase and remove the commit that adds the libs from this PR

coreyhulen commented:
@cpanato can you list the new libs for @crspeller to add?

cpanato commented:
@coreyhulen @crspeller 

package: github.com/jaytaylor/html2text
package: gopkg.in/gomail.v2
   version: 2.0.0

coreyhulen commented:
awesome...thanks.

jasonblais commented:
I've added a WIP label until the libraries are added to `master`

cpanato commented:
then after that I will remove the commit from the PR and rebase to get the libs

cpanato commented:
@jasonblais let me know when I can update this branch

jasonblais commented:
Hey @cpanato,

looks like #5748 was merged so you can pull the updates from `master` and we can test it out

cpanato commented:
@jasonblais in the build I'm getting this:
```
vendor/github.com/jaytaylor/html2text/html2text.go:10:2: cannot find package "github.com/ssor/bom" in any of:
	/home/ubuntu/workspace/mattermost-platform-pr/src/github.com/mattermost/platform/vendor/github.com/ssor/bom (vendor tree)
	/usr/local/go/src/github.com/ssor/bom (from $GOROOT)
	/home/ubuntu/workspace/mattermost-platform-pr/src/github.com/ssor/bom (from $GOPATH)
```
we need to add that lib as well ?

cpanato commented:
Ok, rebased and I think is good for review :)

jasonblais commented on glide.yaml:
```diff
@@ -79,3 +79,4 @@ import:
 - package: github.com/jaytaylor/html2text
 - package: gopkg.in/gomail.v2
   version: 2.0.0
+- package: github.com/ssor/bom
```
Looks like this project doesn't have a license yet -- I've opened an issue asking if they'd be open to add an MIT or Apache v2 license for the repo: https://github.com/ssor/bom/issues/1

jasonblais replied:
@cpanato I haven't heard a response yet about adding an open-source license -- while we should wait for a few more days, I'm wondering if there are other alternatives we could consider?

cpanato replied:
I will take a look if have other alternatives

cpanato replied:
I can fork the html2text and remove the bom library and write my own without using external lib.

jasonblais replied:
Sounds good @cpanato

esethna commented:
Thanks for another great contribution @cpanato, your work here is much appreciated! Just letting you know that there may be a delay in reviewing and merging this PR since we are approaching code complete for v3.8, releasing April 16th. Please see our [release process](https://docs.mattermost.com/process/release-process.html) for more details.

cpanato commented:
Thanks @esethna, this PR needs some refactor due a missing license in one lib i used, then I will write one to change that.

cpanato commented:
Set the Title to WIP in order to refactor the library that converts html to text.
is that good @jasonblais ?

jasonblais commented:
yep, sounds good! Also added 'Work In Progress' label for now

cpanato commented:
As we discussed @jasonblais. Please review with @crspeller this updated libs, and if need I can drop the commit and wait for the library be added in another PR.

cpanato commented:
any update @jasonblais ?

jasonblais commented:
@cpanato Sorry, this slipped my mind -- can you confirm a list of added and removed libraries by this PR?

cpanato commented:
@jasonblais 

added: package: github.com/cpanato/html2text
removed: github.com/jaytaylor/html2text

the issue here, is when I added the lib it update some others.

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

jasonblais commented:
Yep, just wanted to confirm those were the libraries -- I'll ask Chris to include them in his dependency update, which he'll submit soon.

In the meantime, we can start testing it, we don't have to wait on PM review. @lindalumitchell would you be able to help test when you get the chance?

mattermod commented:
Spinmint test server created at: http://i-00a7d2bad4948a70a.spinmint.com:8065/pr5698

Test Account 1: Email: `test@test.com` | Password: `passwd`

Test Account 2: Email: `test2@test.com` | Password: `passwd`

Instance ID: i-00a7d2bad4948a70a

lindalumitchell commented:
@jasonblais @cpanato I tested invite and notification emails from the spinmint, on Thunderbird defaulting to plain text, and viewed original in Gmail to see multipart. All URLs, markdown, emoji, permalinks that I tested behaved as expected. No issues found.

jasonblais commented:
Thanks! Also did a smoke test of email notifications in HTML and looks as expected.

@cpanato I'll keep this in PM review until Chris has submitted a PR that updates dev dependancies ([I asked him to add cpanato/html2text and remove jaytaylor/html2](https://pre-release.mattermost.com/core/pl/1mexdqwiy7ggukf1rp4noq7rgc)). The PR should come soon given 3.8.0 is out now

cpanato commented:
@jasonblais cool, I will wait for this PR and then Update this branch.

jasonblais commented:
here's the PR @cpanato https://github.com/mattermost/platform/pull/6215

cpanato commented:
@jasonblais there is one problem, this lib was removed and should not 
```
-- package: gopkg.in/gomail.v2
 -  version: 2.0.0
```

jasonblais commented:
@cpanato https://github.com/mattermost/platform/pull/6431 containing your libraries is now merged.

Please help rebase your branch and I'll queue for dev review?

Thanks for your patience here.

cpanato commented:
@jasonblais rebased

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: http://i-0041f8753f1f49c33.spinmint.com:8065/pr5698

Test Account 1: Email: `test@test.com` | Password: `passwd`

Test Account 2: Email: `test2@test.com` | Password: `passwd`

Instance ID: i-0041f8753f1f49c33

lindalumitchell commented:
@jasonblais I'm unable to get email working on the spinmint. I enter the same SMTP info as a CI server or bladekick, and when I test the connection it fails ("Connection unsuccessful: Failed to set HELO - smtp: Hello called after other methods"). This has worked in the past for me on other spinmints. 

Any ideas? I've tried different Connection Security settings and played with toggling other options as well, and no luck. Would love help figuring out if it's my error or a real issue.

cpanato commented:
@lindalumitchell @jasonblais this is might be related to the https://github.com/mattermost/platform/pull/6429 which add the `HELO` before connect to the SMTP. But looks like we have some issues if the SiteUrl is not the same for the hostname. I think we need to add @hmhealey to we discuss this `HELO`

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: http://i-05fd12b9952a911e6.spinmint.com:8065/pr5698

Test Account 1: Email: `test@test.com` | Password: `passwd`

Test Account 2: Email: `test2@test.com` | Password: `passwd`

Instance ID: i-05fd12b9952a911e6

cpanato commented:
@lindalumitchell rebased to get the fix for the HELO, can you please spin a new test server?

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: http://i-0b968d73c9fb5084c.spinmint.com:8065/pr5698

Test Account 1: Email: `test@test.com` | Password: `passwd`

Test Account 2: Email: `test2@test.com` | Password: `passwd`

Instance ID: i-0b968d73c9fb5084c

jasonblais commented:
@lindalumitchell This one is ready for testing, the email setup issue should now be fixed

jasonblais commented:
FYI @cpanato, @lindalumitchell  I'll need to set up a mailjet account for spinmints as we block SES credentials by IP addresses.

I'll try to look into this tomorrow. Hopefully we can get this to dev review soon, sorry for the drag here Carlos

cpanato commented:
@jasonblais any update? :)

jasonblais commented:
@lindalumitchell When you have the time, you can configure spinmints to send emails via the credentials at https://github.com/mattermost/platform-private/blob/master/config/spinmint-email.txt

Please make sure to change the default admin passwords before setting this up as usual.

lindalumitchell commented:
@jasonblais @cpanato we got email set up on the spinmint, and it all looks good. As before, I tested invite and notification emails, viewing plain text in Thunderbird and HTML in Gmail. No issues found.

jasonblais commented:
Thanks @lindalumitchell! Really appreciate it

crspeller commented on app/email.go:
```diff
@@ -9,6 +9,7 @@ import (
 	"net/url"
 
 	l4g "github.com/alecthomas/log4go"
+	"github.com/cpanato/html2text"
```
Why are we using a specialized version of the library? Can we avoid this? If not, it should be hosted in the Mattermost org.

cpanato replied:
lets discuss that in the channel, I will ping you

cpanato replied:
discussed in the channel and transferred the repo :)

crspeller commented on app/email.go:
```diff
@@ -26,7 +27,10 @@ func SendChangeUsernameEmail(oldUsername, newUsername, email, locale, siteURL st
 		map[string]interface{}{"TeamDisplayName": utils.Cfg.TeamSettings.SiteName, "NewUsername": newUsername}))
 
-	if err := utils.SendMail(email, subject, bodyPage.Render()); err != nil {
+	renderedHtmlBody := bodyPage.Render()
```
Maybe there should be a function that takes a bodyPage and does all this and send the mail. So this doesn't have to be copy coded everywhere.

cpanato replied:
done

cpanato commented:
@jasonblais @crspeller the build now will fail because need to update the glide to ude the mattermost/html2text repo.

jasonblais commented:
@crspeller Are we going to include https://github.com/mattermost/html2text in the next dependancy update?

jasonblais commented:
Chatted with @crspeller - we'll include https://github.com/mattermost/html2text in the dependancy update for 4.2. Unfortunately it missed the one for 4.1

cpanato commented:
I will wait for the dependencies update then after that will rebase and fix the conflict, is that ok @jasonblais?

jasonblais commented:
Sounds good, thanks! Hopefully we'll get it into 4.2 :)

cpanato commented:
we can make a party 🎉 :)

cpanato commented:
@jasonblais any news on this?

jasonblais commented:
The next dependancy update should be coming next week

crspeller commented:
@jasonblais @cpanato The dependency update has gone in for this.

cpanato commented:
@jasonblais rebased!

jasonblais commented:
@crspeller please help merge the PR for 4.2
